src/xdp-app-info.c

```diff
@@ -742,35 +755,60 @@ on_peer_died (const char *name)
 }
 
 static XdpAppInfo *
-xdp_connection_lookup_app_info_sync (GDBusConnection  *connection,
+maybe_create_test_app_info (void)
+{
+  const char *test_override_app_id;
+  const char *test_override_usb_queries;
+
+  test_override_app_id = g_getenv ("XDG_DESKTOP_PORTAL_TEST_APP_ID");
+  if (!test_override_app_id)
+    return NULL;
+
+  test_override_usb_queries = g_getenv ("XDG_DESKTOP_PORTAL_TEST_USB_QUERIES");
+  return xdp_app_info_test_new (test_override_app_id,
+                                test_override_usb_queries);
+}
+
+static XdpAppInfo *
+maybe_create_registered_test_app_info (const char *registered_app_id)
+{
+  const char *test_override_app_id;
+  const char *test_override_usb_queries;
+
+  test_override_app_id = g_getenv ("XDG_DESKTOP_PORTAL_TEST_APP_ID");
+  if (!test_override_app_id)
+    return NULL;
+
+  test_override_usb_queries = g_getenv ("XDG_DESKTOP_PORTAL_TEST_USB_QUERIES");
+  return xdp_app_info_test_new (registered_app_id,
+                                test_override_usb_queries);
+}
+
+static XdpAppInfo *
+xdp_connection_create_app_info_sync (GDBusConnection  *connection,
                                      const char       *sender,
                                      GCancellable     *cancellable,
                                      GError          **error)
 {
   g_autoptr(XdpAppInfo) app_info = NULL;
   g_autofd int pidfd = -1;
   uint32_t pid;
-  const char *test_override_app_id;
-  const char *test_override_usb_queries;
   g_autoptr(GError) local_error = NULL;
-
-  app_info = cache_lookup_app_info_by_sender (sender);
-  if (app_info)
-    return g_steal_pointer (&app_info);
+  const char *app_info_kind = NULL;
 
   if (!xdp_connection_get_pidfd (connection, sender, cancellable, &pidfd, &pid, error))
     return NULL;
 
-  test_override_app_id = g_getenv ("XDG_DESKTOP_PORTAL_TEST_APP_ID");
-  test_override_usb_queries = g_getenv ("XDG_DESKTOP_PORTAL_TEST_USB_QUERIES");
-  if (test_override_app_id)
-    {
-      app_info = xdp_app_info_test_new (test_override_app_id,
-                                        test_override_usb_queries);
-    }
+  app_info = maybe_create_test_app_info ();
+  if (app_info)
+    app_info_kind = "test";
 
   if (app_info == NULL)
-    app_info = xdp_app_info_flatpak_new (pid, pidfd, &local_error);
+    {
+      app_info = xdp_app_info_flatpak_new (pid, pidfd, &local_error);
+      if (app_info)
+        app_info_kind = "flatpak";
```

Comment:
```suggestion
      app_info_kind = "flatpak";
```

Reply:
Creating `app_info` might have failed, and don't want to set it if it did, even if it'd be overwritten eventually.

Reply:
The other ones do exactly that though, so this isn't consistent.

Reply:
What other ones do you mean?

Reply:
`app_info_kind = "snap";` and `app_info_kind = "derived host";` below

Reply:
Ah, indeed. The intention was to make those who conditionally work conditionally set it, but seems I did the wrong thing for the snap case.

Reply:
On the other hand, not conditioning them ends up with intermediate incorrect state, which I found misleading.